### ISSUE TITLE
Static files not working when not in DEBUG=True mode

### ISSUE THREAD

**Issue by timurjavid** [uiuc-mrdc/cogs#6]
We need to figure out how to serve static files correctly in production. Running in DEBUG=True mode is insecure. Might have to tweak views.py and settings.py.

**Comment by mgale456** [comment]
Only serving static files while DEBUG=true seems to be an intended feature. Once we start thinking about moving to production it should be no big deal to put the static files in the correct directory to be served by the server, and not through Django.

Django's built in testserver is for convenient development testing, and not intended for deployment.

I'll leave this issue open, since it is still an unsolved issue, but not an urgent issue

**Comment by lpulley** [comment]
I'm not sure what that means precisely. What will be serving the files in this case? You mean like a static file server in Nginx or similar? I could whip that together really, really quickly. Let's aim to get this solved by the end of Saturday's meeting, at the latest.

**Comment by mgale456** [comment]
Yes. According to the Django tutorial and Django's [deployment documentation](https://docs.djangoproject.com/en/2.2/howto/deployment/), Django is run in production through Apache, Nginx, or your favorite webserver compatible with WSGI. The built in server we're running is for development and testing and is not secure or efficient.

It is important to note that Django Channels necessitates a server compatible with ASGI, and there are a lot fewer options. Primarily `Daphne` which is developed specifically for Channels.

**Comment by lpulley** [comment]
Interesting... I'll look into Daphne.

**Comment by mgale456** [comment]
Reviving this Issue. Very relevant.

I have made a `deployment` branch. `Daphne` works, and handles the HTTP and Websocket connections, however it does **not** serve any static files (css, icons, etc.). This is by design for `Daphne`, not a bug.

We need to set up something else like `nginx` to serve those, while Daphne handles the Django requests. More details are available, as well as an example at the bottom that doesn't use Docker on the [Channels Docs for Deployment](https://channels.readthedocs.io/en/latest/deploying.html).

Other notes:
Django has a command `collectstatic` that copies all of the static files into a single directory from all of the apps and such. It is configured at the bottom of `settings.py` to put everything in cogs/django/nginx/ on the host.

**Comment by lpulley** [comment]
That sounds not too hard. I'm busy the next few days but I could do this.

**Comment by lpulley** [comment]
Hello again 👋 

I've got a lot of experience with this now. I'm confident I can get this sorted out. Drop Daphne, grab Uvicorn and Gunicorn and set the container command to `gunicorn example.asgi:application --bind 0.0.0.0:8000 -w 4 -k uvicorn.workers.UvicornWorker`. For automatic code reloading in development, override that command (in `docker-compose.yml`) to `uvicorn example.asgi:application --host 0.0.0.0 --reload` (doesn't do multiple workers, but does do code reloading).

When I have time to work on this, I can take care of this.

**Comment by mgale456** [comment]
It's not too hard in Daphne either. Just set up an nginx reverse proxy to
serve files and send page requests to Daphne. I have it set up on my
personal project, but I'd guess guvicorn may have other pros since your
company chose it.

**Comment by mgale456** [comment]
Another note supporting Uvicorn/Gunicorn. If they work identically easily in dev (with reloading) and production, that is awesome. It is a bit annoying to switch to daphne, and then sometimes not get the same results if you're not serving the files quite right in nginx and there are some edge cases my nginx implementation does not cover that I wouldn't expect others to know about so much.

**Comment by lpulley** [comment]
For dev, just change the command in compose. The Dockerfile command should be gunicorn (which doesn't really do good reloading, but does do workers) and then dev should use plain uvicorn (which does do good reloading)

**Comment by lpulley** [comment]
I have a whole setup for this that I learned from research. It's very nice
